### ISSUE TITLE
cppcheck: catch ints as bools

### ISSUE THREAD

**Issue by adeebshihadeh** [commaai/panda#2105]
Context in #1876

This is three steps:
1. add support for this to cppcheck
2. upstream that support to cppcheck
3. PR to enable it and fix all the violations

While we're waiting for them to merge the support upstream, you can open a draft PR using the patch.

**Comment by anshgoyalevil** [comment]
Hey @adeebshihadeh I am willing to resolve this issue. Can you please assign it to me?

**Comment by anshgoyalevil** [comment]
I have created a pull request for getting the required changes merged into upstream cppcheck. Lets see how it goes.
https://github.com/danmar/cppcheck/pull/7110

**Comment by eckertliam** [comment]
@adeebshihadeh it is unclear to me whether this issue is still open/active. If it is I am willing to resolve this if you assign it to me.

**Comment by B-L-A-N-K-official** [comment]
is this open ?

**Comment by varshneydevansh** [comment]
@rbxict I am workign on this and that my upstream merge 
still have a bug which I found after runnning my changes.
and I have found the places which needs to be fix in the panda :)

**Comment by varshneydevansh** [comment]
@adeebshihadeh I identified the remaining panda-side cppcheck cleanup after 
the upstream fixes

- https://github.com/danmar/cppcheck/pull/8340
- https://github.com/danmar/cppcheck/pull/8388 (still pending)

 and already made the change here:

https://github.com/varshneydevansh/panda/commit/286147eedc9da60a14228f470a4c42a1704d39d5

I’m blocked by GitHub fork/PR creation issues on `commaai/panda`. 
**If needed, could you open/apply this tiny panda patch on my behalf?**




```sh

 Devanshs-MacBook-Pro    \panda  ➜ (  zero_and_one)   25ms   3:55 PM   
 ⚡devanshvarshney ❯❯ source .venv/bin/activate
SKIP_BUILD=1 SKIP_TABLES_DIFF=1 CPPCHECK_DIR=/Users/devanshvarshney/cppcheck tests/misra/test_misra.sh


** PANDA H7 CODE **
Checking /Users/devanshvarshney/panda/board/main.c ...
/Users/devanshvarshney/panda/board/main.c:23:2: information: Include file: "board/obj/gitversion.h" not found. [missingInclude]
#include "board/obj/gitversion.h"
 ^
Checking /Users/devanshvarshney/panda/board/main.c: __GNUC__=9;STM32H7=1;STM32H725xx=1...
/Users/devanshvarshney/panda/board/stm32h7/llspi.h:2:30: style: Parameter 'addr' can be declared as pointer to const [constParameterPointer]
void llspi_mosi_dma(uint8_t *addr, int len) {
                             ^
/Users/devanshvarshney/panda/board/stm32h7/llspi.h:29:30: style: Parameter 'addr' can be declared as pointer to const [constParameterPointer]
void llspi_miso_dma(uint8_t *addr, int len) {
                             ^
/Users/devanshvarshney/panda/.venv/lib/python3.12/site-packages/opendbc/safety/modes/toyota.h:334:10: style: misra violation (use --rule-texts=<file> to get proper output) [misra-c2012-10.3]
      tx = 0;
         ^
/Users/devanshvarshney/panda/board/drivers/can_common.h:44:12: style: misra violation (use --rule-texts=<file> to get proper output) [misra-c2012-10.3]
  bool ret = 0;
           ^
/Users/devanshvarshney/panda/board/drivers/can_common.h:54:9: style: misra violation (use --rule-texts=<file> to get proper output) [misra-c2012-10.3]
    ret = 1;
        ^
/Users/devanshvarshney/panda/board/drivers/gpio.h:16:15: style: misra violation (use --rule-texts=<file> to get proper output) [misra-c2012-12.2]
  tmp &= ~(3U << (pin * 2U));

```

```sh
 Devanshs-MacBook-Pro    \panda  ➜ (  zero_and_one)   4.542s   5:58 PM   
 ⚡devanshvarshney ❯❯ gh pr create \
  --repo commaai/panda \
  --base master \
  --head varshneydevansh:zero_and_one

Creating pull request for varshneydevansh:zero_and_one into master in commaai/panda

? Title (required) cppcheck: replace bool 0/1 assignments with false/true
? Body <Received>
? What's next? Continue in browser
Opening https://github.com/commaai/panda/compare/master...varshneydevansh:zero_and_one in your browser.
```

------



<img width="806" height="677" alt="Image" src="https://github.com/user-attachments/assets/b6227bb4-ed15-4cc5-8616-f569d0ce0d65" />


<img width="1404" height="857" alt="Image" src="https://github.com/user-attachments/assets/901a5f53-219d-4737-9279-6165b52f8537" />

**Comment by varshneydevansh** [comment]
I am able to fork now and I am waiting for the reamining regression 
approval in the upstream once it is done I will open the PR for this :)

**Comment by varshneydevansh** [comment]
Also wanted to clarify the expected workflow for completing this 
bounty now that the upstream cppcheck changes are merged.

The remaining step seems to be updating  [`commaai/dependencies@release-cppcheck`](https://github.com/commaai/dependencies/tree/master/cppcheck) 
so panda uses the cppcheck version containing:

- https://github.com/cppcheck-opensource/cppcheck/pull/8340
- https://github.com/cppcheck-opensource/cppcheck/pull/8388

I’m holding off on opening the PR until the dependency update lands.

Is that the right order, or should I open the panda PR now as a 
draft while the dependency update is pending?


`panda` currently uses:

```bash
Collecting cppcheck@ git+https://github.com/commaai/dependencies.git@release-cppcheck#subdirectory=cppcheck (from pandacan==0.0.10)
```

**Comment by adeebshihadeh** [comment]
Oh nice, just open a PR to dependencies and here, then get both green. You can just point your panda branch to your dependencies branch.

**Comment by varshneydevansh** [comment]
The upstream cppcheck fixes are merged, but the latest cppcheck 
release available to package still predate the second fix (`e41115d` 
from cppcheck-opensource/cppcheck#8388) which is `2.20.0`.

I’ll wait for the next cppcheck release that includes `e41115d`, then 
open the dependencies PR and update the panda PR to use it.

**Comment by varshneydevansh** [comment]
I talked with Daniel and he said that we should
wait for a while for the new relaese of cppcheck 
2.21.0 version.

**Comment by varshneydevansh** [comment]
Hi @adeebshihadeh ,

I have opened the PR to https://github.com/commaai/dependencies/pull/85 as asked & 
also one to [openbdc](https://github.com/commaai/opendbc/pull/3444) as I found this being flagged by cppcheck.

I have tested carefully with my merged upstream cppcheck release
and only 2 places were flagged https://github.com/commaai/panda/pull/2409

**Comment by varshneydevansh** [comment]
Maybe now we can mark this as complete :)
